posthog/models/insight.py

```diff
@@ -34,9 +35,9 @@ class Insight(models.Model):
     derived_name: models.CharField = models.CharField(max_length=400, null=True, blank=True)
     description: models.CharField = models.CharField(max_length=400, null=True, blank=True)
     team: models.ForeignKey = models.ForeignKey("Team", on_delete=models.CASCADE)
-    filters: models.JSONField = models.JSONField(default=dict)
+    _filters: models.JSONField = models.JSONField(default=dict, db_column="filters")
```

Comment:
I can see this is still draft, but wanted to chime in here because I'd want to follow whatever pattern you decide here for experiments / anything else we have currently using filters.

It's not yet clear to me what benefit renaming here gives us, as it means refactoring a lot of annoying ORM level queries.

I think I'd prefer having an explicit `.get_query_or_filter` function or parameter that returns the query if it exists, and if not, generates it from the filter. And then, use this wherever we need to input the query. This way old test setup everywhere + ORM interaction just works, and we can be sure that we're using the right thing for results by asserting whether get_query_or_filter was called or not.

Reply:
oh with this I also don't have to worry about the API interface breaking, since I'd expect all API clients would want filters to work coming from the API, even if internally we then convert it to a query.

Reply:
> It's not yet clear to me what benefit renaming here gives us, as it means refactoring a lot of annoying ORM level queries.
> 
> I think I'd prefer having an explicit `.get_query_or_filter` function or parameter that returns the query if it exists, and if not, generates it from the filter. And then, use this wherever we need to input the query. This way old test setup everywhere + ORM interaction just works, and we can be sure that we're using the right thing for results by asserting whether get_query_or_filter was called or not.

So basically the goal of this PR is that when the flag is enabled, we will only have a `query` frontend side, but never `filters`. That way we can be sure we aren't relying on filters frontend side any more. I think there are two places where I could implement this functionality: (a) in the DRF serializer for insights (b) in the django model. I went for (b) as I thought this might be easier.

The renaming is necessary, as otherwise Django does not allow to have a getter/setter for an existing field. Seems to be a long standing issue, but I'm not Django expert so feedback on how else I can achieve this is welcome. I'll try the DRF serializer approach to see if this easier.

Reply:
I see - thanks for explaining the reasoning & goals. So effectively the goal is to never return filters to the frontend if the flag is enabled. Hmm, I see.

If you only want this at read time, and even with the flag on someone can send filters to create an insight (but we don't return filters - which seems reasonable as API users don't necessarily get updated with the flag), then I think you can achieve this very cleanly using a SerializerMethodField -> https://www.django-rest-framework.org/api-guide/fields/#serializermethodfield - just convert `filters` and `query` to write-only in the model serializer, and add 2 readonly methods for `filters` and `query` which basically do what you're doing. This should isolate the change only at the serializer level.

I don't really have strong feelings anymore, since it seems like experiments don't have to follow this convention, we'll just go all or nothing once this test phase is done 👍